src/self_driving/pilot_util.cpp

```diff
@@ -12,19 +12,149 @@
+    txn_manager->Commit(txn, transaction::TransactionUtil::EmptyCallback, nullptr);
+  }
+}
+
+std::unique_ptr<planner::AbstractPlanNode> PilotUtil::GetOutPlan(
+    transaction::TransactionContext *txn, common::ManagedPointer<catalog::CatalogAccessor> accessor,
+    common::ManagedPointer<std::vector<parser::ConstantValueExpression>> params,
+    common::ManagedPointer<std::vector<type::TypeId>> param_types,
+    common::ManagedPointer<parser::ParseResult> stmt_list, catalog::db_oid_t db_oid,
+    common::ManagedPointer<optimizer::StatsStorage> stats_storage, uint64_t optimizer_timeout) {
+  auto binder = binder::BindNodeVisitor(common::ManagedPointer(accessor), db_oid);
+  binder.BindNameToNode(stmt_list, params, param_types);
+  // Creating exec_ctx
+  std::unique_ptr<optimizer::AbstractCostModel> cost_model = std::make_unique<optimizer::TrivialCostModel>();
+
+  auto out_plan = trafficcop::TrafficCopUtil::Optimize(common::ManagedPointer(txn), common::ManagedPointer(accessor),
+                                                       common::ManagedPointer(stmt_list), db_oid, stats_storage,
+                                                       std::move(cost_model), optimizer_timeout)
+                      ->TakePlanNodeOwnership();
+  return out_plan;
+}
+
+void PilotUtil::GetQueryPlans(common::ManagedPointer<Pilot> pilot, common::ManagedPointer<WorkloadForecast> forecast,
+                              uint64_t end_segment_index,
+                              std::vector<std::unique_ptr<planner::AbstractPlanNode>> *plan_vecs) {
+  auto txn_manager = pilot->txn_manager_;
+  auto catalog = pilot->catalog_;
+  transaction::TransactionContext *txn = txn_manager->BeginTransaction();
+
+  for (auto idx = 0; idx <= end_segment_index; idx++) {
+    for (auto &it : forecast->forecast_segments_[idx].id_to_num_exec_) {
+      auto stmt_list = parser::PostgresParser::BuildParseTree(forecast->query_id_to_text_[it.first]);
+      auto db_oid = static_cast<catalog::db_oid_t>(forecast->query_id_to_dboid_[it.first]);
+      auto accessor = catalog->GetAccessor(common::ManagedPointer(txn), db_oid, DISABLED);
+
+      auto out_plan = PilotUtil::GetOutPlan(
+          txn, common::ManagedPointer(accessor), common::ManagedPointer(&forecast->query_id_to_params_[it.first][0]),
+          common::ManagedPointer(&(forecast->query_id_to_param_types_[it.first])), common::ManagedPointer(stmt_list),
+          db_oid, pilot->stats_storage_, forecast->optimizer_timeout_);
+
+      plan_vecs->emplace_back(std::move(out_plan));
+    }
+  }
+  txn_manager->Abort(txn);
+}
+
+uint64_t PilotUtil::ComputeCost(common::ManagedPointer<Pilot> pilot, common::ManagedPointer<WorkloadForecast> forecast,
+                                uint64_t start_segment_index, uint64_t end_segment_index) {
+  // Compute cost as average latency of queries weighted by their num of exec
+  // Keep track of the ou inference results (a vector of doubles) for each ous associated with the pipeline
```

Comment:
nit: First explain the map first: maps *** to ***. Then explain the meaning of the value (the vector). Otherwise, it may be a bit confusing.